kvserverpb: move quorum safeguard into execChangeReplicasTxn

tbg opened the pull request:
This used to live in the replicate queue, but there are other
entry points to replication changes, notably the store rebalancer
which caused #54444.

Move the check in the guts of replication changes where it is
guaranteed to be invoked.

Fixes #50729
Touches #54444 (release-20.2)

@aayushshah15 only requesting your review since you're in the area.
Feel free to opt out.

Release note (bug fix): in rare situations, an automated replication
change could result in a loss of quorum. This would require down nodes
and a simultaneous change in the replication factor. Note that a change
in the replication factor can occur automatically if the cluster is
comprised of less than five available nodes. Experimentally the likeli-
hood of encountering this issue, even under contrived conditions, was
small.

tbg committed:
kvserverpb: move quorum safeguard into execChangeReplicasTxn

This used to live in the replicate queue, but there are other
entry points to replication changes, notably the store rebalancer
which caused #54444.

Move the check in the guts of replication changes where it is
guaranteed to be invoked.

Fixes #50729
Touches #54444 (release-20.2)

Release note (bug fix): in rare situations, an automated replication
change could result in a loss of quorum. This would require down nodes
and a simultaneous change in the replication factor. Note that a change
in the replication factor can occur automatically if the cluster is
comprised of less than five available nodes. Experimentally the likeli-
hood of encountering this issue, even under contrived conditions, was
small.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56735)
<!-- Reviewable:end -->

tbg commented:
I'm running 100 iterations of `replicate/wide` on this change. This test didn't previously replicate the issue (save on certain commits), but since I moved the protection that that test *does* exercise deeper down, I will be confident that absent any problems I've fixed it.

tbg commented:
They all passed.

tbg commented:
@andreimatei you might be interested in this since you just touched `execChangeReplicasTxn`. (I'll deal with your merge conflict later).

aayushshah15 reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56735#-:-MMTG8ec5Gh511lv8-f9:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15 and @knz)



<!-- Sent from Reviewable.io -->

tbg commented:
Moving the check is causing some new flakes due to getting different errors from replication changes, I'll take the opportunity to give this an appropriate `cockroachdb/errors` treatment so that these tests never have to string match again.

andreimatei reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56735#-:-MMWa4RL1xN9zCcW-0bW:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

tbg commented:
Ready for another review pass now. I went ahead and used `errors.Mark` for all retriable conditions that can come out of replication changes..

tbg commented:
Ok, fixed one last issue, CI should be green now.

andreimatei commented:
don't forget to squash

tbg committed:
kvserver: fold two tests into subtests of one

Release note: None

tbg commented:
@aayushshah15 I'm running into an awkward flake in TestReplicateAddAndRemove (but it's a general problem). I'm running a manual upreplication but the replicate queue sees the learner and decides to remove it. The test's upreplication retries and succeeds (it's smart enough to let the replicate queue finish if it decides to add as well), but now we've given out a different replicaID than the test expects.

I don't know why this happens only as of this PR, but I guess I did change something about how aggressively the replicate queue retries. My assumption is that it's marking the "cannot add replica that is already in descriptor" errors as retryable in `validateReplicationChanges` is what did us in.

Since you've been in the area recently, do you have any suggestions for fixing this? I can disable the replicate queue in this test (that it passed before was more luck than design), but I imagine this is a general pattern that has been annoying for you as well.

aayushshah15 commented:
I ran into this issue on one of my changes before, and as you mentioned, disabling the replicate queue does cure the flakes. I'm actually super surprised that this test wasn't already flaking before though. 

> My assumption is that it's marking the "cannot add replica that is already in descriptor" errors

Out of curiosity, why are we now marking these as retryable? The error message makes me think that those shouldn't be retried.

tbg commented:
Note to self: I should massage that error matching grab bag method touched in https://github.com/cockroachdb/cockroach/pull/57272 as well.

tbg commented:
> Out of curiosity, why are we now marking these as retryable? The error message makes me think that those shouldn't be retried.

You're absolutely right, I simply wasn't paying attention. I fixed this - the flake remains. I think it's just that the genuine race that was always present now actually kicks in in these tests thanks to different timing. I'll fix the by disabling the replicate queue. Maybe one day we will start using advisory locks on the range descriptor for these multi-step operations, but not anytime soon.

tbg committed:
kvserver: improve multiTestContext.changeReplicas

The "already done" code wasn't working any more.

Release note: None

tbg commented:
`make stress PKG=./pkg/kv/kvserver/ STRESSFLAGS='-p 4' TESTFLAGS='-show-logs'` finally looks stable enough:

```
36 runs so far, 0 failures, over 32m30s
```

tbg commented:
bors r=aayushshah15

andreimatei commented:
mmm don't know about stable :P
The first commit makes `make stress PKG='./pkg/kv/kvserver' TESTS=TestChangeReplicasDescriptorInvariant` fail real fast on GCE. I think it's a manual replication vs queue race as was discussed above. I could disable the queue in that test, but how many others are there? It's also not clear to me from the discussion whether we understand why this patch makes the race more likely. Tobi, do you mind looking into it?
And btw, if this is the time that we make the race impossible by somehow giving the manual replication some grace period after adding a learner to promote it, someone gets my peer ack.

andreimatei commented:
Same with `TestReplicateRange`.

tbg commented:
I'll take a look today.

On Fri, Dec 4, 2020, 01:37 Andrei Matei <notifications@github.com> wrote:

> Same with TestReplicateRange.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/56735#issuecomment-738473681>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABGXPZAODEAGQGV5SYK27PTSTAVMZANCNFSM4TXIPXHQ>
> .
>

tbg commented:
Ok, I understand this. Adding the check back in the replicate queue "fixes" the problem. It's because in its new location down in the voter promotion, the check simply fires late. Essentially the following happens:

- manual replication adds learner
- manual replication tries to add voter, but gets held up by the check, goes to sleep
- queue tries to upreplicate, sees learner, removes learner
- boom

Previously, we'd have the following

- queue is blocked by the check
- manual replication adds learner
- manual replication adds voter
- queue gets past the check

So essentially the fact that it takes a split second for everyone to seem live at the beginning of the cluster was ensuring that early replication changes would never contend with the replicate queue's.
